app.py

```diff
@@ -0,0 +1,64 @@
+import os
+
+from flask import Flask, render_template, flash, redirect, session, g
+from flask_debugtoolbar import DebugToolbarExtension
+from sqlalchemy.exc import IntegrityError
+
+from models import db, connect_db, User
+from forms import UserAddForm
+
+from dotenv import load_dotenv
+
+load_dotenv()
+
+app = Flask(__name__)
+
+app.config["SQLALCHEMY_DATABASE_URI"] = os.environ.get("DATABASE_URL")
+app.config["SQLALCHEMY_TRACK_MODIFICATIONS"] = False
+app.config["SQLALCHEMY_ECHO"] = False
+app.config["DEBUG_TB_INTERCEPT_REDIRECTS"] = False
+app.config["SECRET_KEY"] = os.environ.get("SECRET_KEY", "shhhItsASecret123")
+toolbar = DebugToolbarExtension(app)
+
+connect_db(app)
+
+
+@app.before_request
+def add_user_to_g():
+    """If logged in, add user to Flask g."""
+    if session.get("user_id") is not None:
+        g.user = User.query.get(session["user_id"])
+
+    else:
+        g.user = None
+
+
+@app.route("/")
+def home():
+    """Render home page."""
+    return render_template("index.html")
+
+
+@app.route("/signup", methods=["GET", "POST"])
+def signup():
+    """Render AddUserForm form and creates a new user."""
+    form = UserAddForm()
+
+    if form.validate_on_submit():
+        try:
+            user = User.signup(
+                username=form.username.data,
+                password=form.password.data,
+                email=form.email.data,
+            )
+            db.session.commit()
+            session["user_id"] = user.id
+            flash(f"Welcome {user.username}!", "success")
+
+            return redirect("/")
+
+        except IntegrityError:
+            flash("Username already taken", "danger")
+            return render_template("signup.html", form=form)
```

Comment:
@blakely the application only handles the IntergrityError. You should be able to handle multiple errors.

Reply:
I realized it wasn't always catching the IntegrityError and I needed to rollback the session to handle the error so it should work for the unique violation now. I'm not sure what other specific errors I should be handling so I added a page for general errors for now. How should I handle error types that I can't anticipate?

Reply:
The ideal way would be to anticipate any error in the `except` block and flash the returned error if any. But that implementation is OK for now.